cloud/linode/instances.go

```diff
@@ -29,176 +27,76 @@ func (e instanceNoIPAddressesError) Error() string {
+	// An instance is considered to be "shutdown" when it is
+	// in the process of shutting down, or already offline.
+	if instance.Status == linodego.InstanceOffline ||
+		instance.Status == linodego.InstanceShuttingDown {
+		return true, nil
 	}
-	return strconv.Itoa(linode.ID), nil
-}
 
-func (i *instances) InstanceType(ctx context.Context, nodeName types.NodeName) (string, error) {
-	ctx = sentry.SetHubOnContext(ctx)
-	sentry.SetTag(ctx, "node_name", string(nodeName))
-
-	linode, err := linodeByName(ctx, i.client, nodeName)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return "", err
-	}
-	return linode.Type, nil
+	return false, nil
 }
 
-func (i *instances) InstanceTypeByProviderID(ctx context.Context, providerID string) (string, error) {
-	ctx = sentry.SetHubOnContext(ctx)
-	sentry.SetTag(ctx, "provider_id", providerID)
-
-	id, err := parseProviderID(providerID)
+func (i *instances) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	linode, err := i.lookupLinode(ctx, node)
 	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return "", err
-	}
-
-	sentry.SetTag(ctx, "linode_id", strconv.Itoa(id))
-
-	linode, err := linodeByID(ctx, i.client, id)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return "", err
+		return nil, err
 	}
-	return linode.Type, nil
-}
 
-func (i *instances) AddSSHKeyToAllInstances(_ context.Context, user string, keyData []byte) error {
-	return cloudprovider.NotImplemented
-}
-
-func (i *instances) CurrentNodeName(_ context.Context, hostname string) (types.NodeName, error) {
-	return types.NodeName(hostname), nil
-}
-
-func (i *instances) InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error) {
-	ctx = sentry.SetHubOnContext(ctx)
-	sentry.SetTag(ctx, "provider_id", providerID)
-
-	id, err := parseProviderID(providerID)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return false, err
+	if len(linode.IPv4) == 0 {
+		return nil, instanceNoIPAddressesError{linode.ID}
 	}
 
-	sentry.SetTag(ctx, "linode_id", strconv.Itoa(id))
+	addresses := []v1.NodeAddress{{Type: v1.NodeHostName, Address: linode.Label}}
 
-	_, err = linodeByID(ctx, i.client, id)
-	if err != nil {
-		if apiError, ok := err.(*linodego.Error); ok && apiError.Code == http.StatusNotFound {
-			return false, nil
+	for _, ip := range linode.IPv4 {
+		ipType := v1.NodeExternalIP
+		if ip.IsPrivate() {
+			ipType = v1.NodeInternalIP
 		}
-		sentry.CaptureError(ctx, err)
-		return false, err
+		addresses = append(addresses, v1.NodeAddress{Type: ipType, Address: ip.String()})
 	}
 
-	return true, nil
-}
-
-func (i *instances) InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error) {
-	ctx = sentry.SetHubOnContext(ctx)
-	sentry.SetTag(ctx, "provider_id", providerID)
-
-	id, err := parseProviderID(providerID)
-	if err != nil {
-		sentry.CaptureError(ctx, err)
-		return false, err
+	// note that Zone is omitted as it's not a thing in Linode
```

Comment:
With Zone being removed, this comment mostly becomes historical.

Reply:
`zones.go` is begin removed, but its functionality is folded into this big metadata method. There really isn't a difference in functionality between master and this PR, the original zones.go code also omitted Zone (as in the struct field), I'm just stating it here explicitly, because it's the only struct field I'm not filling in - so I'm stating, for _future_ reference, that the omission is intentional.